packages/axelar-wasm-std/src/msg_id/base_58_solana_event_index.rs

```diff
@@ -0,0 +1,362 @@
+    }
+
+    pub fn new(tx_id: impl Into<RawSignature>, event_index: impl Into<u32>) -> Self {
+        Self {
+            signature: tx_id.into(),
+            event_index: event_index.into(),
+        }
+    }
+
+    pub fn new_from_b58_encoded_signature_and_index(
+        signature: &str,
+        event_index: impl Into<u32>,
+    ) -> Result<Self, Report<Error>> {
+        Ok(Self {
+            signature: decode_b58_signature(signature)?,
+            event_index: event_index.into(),
+        })
+    }
+}
+
+fn decode_b58_signature(signature: &str) -> Result<RawSignature, Report<Error>> {
+    Ok(bs58::decode(signature)
+        .into_vec()
+        .change_context(Error::InvalidTxDigest(signature.to_string()))?
+        .as_slice()
+        .try_into()
+        .map_err(|e: TryFromSliceError| {
+            Error::InvalidTxDigest(format!(
+                "Signature - {}: {}",
+                signature,
+                e
+            ))
+        })?)
+}
+
+const PATTERN: &str = "^([1-9A-HJ-NP-Za-km-z]{32,88})-(0|[1-9][0-9]*)$";
+lazy_static! {
+    static ref REGEX: Regex = Regex::new(PATTERN).expect("invalid regex");
+}
+
+impl FromStr for Base58SolanaTxDigestAndEventIndex {
+    type Err = Report<Error>;
+
+    fn from_str(message_id: &str) -> Result<Self, Self::Err>
+    where
+        Self: Sized,
+    {
+        // the PATTERN has exactly two capture groups, so the groups can be extracted safely
+        let (_, [signature, event_index]) = REGEX
+            .captures(message_id)
+            .ok_or(Error::InvalidMessageID {
+                id: message_id.to_string(),
+                expected_format: PATTERN.to_string(),
+            })?
+            .extract();
+
+        Ok(Base58SolanaTxDigestAndEventIndex {
+            signature: decode_b58_signature(signature)?,
+            event_index: event_index
+                .parse()
+                .map_err(|_| Error::EventIndexOverflow(message_id.to_string()))?,
+        })
+    }
+}
+
+impl Display for Base58SolanaTxDigestAndEventIndex {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}-{}",
+            bs58::encode(self.signature).into_string(),
+            self.event_index
+        )
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use hex::ToHex;
+
+    use super::*;
+
+    fn random_bytes() -> [u8; 64] {
+        let mut bytes = [0; 64];
+        for b in &mut bytes {
+            *b = rand::random();
+        }
+        bytes
+    }
+
+    fn random_tx_digest() -> String {
+        bs58::encode(random_bytes()).into_string()
+    }
+
+    fn random_event_index() -> u32 {
+        rand::random()
+    }
+
+    #[test]
+    fn should_parse_msg_id() {
+        let res = Base58SolanaTxDigestAndEventIndex::from_str(
+            "4hHzKKdpXH2QMB5Jm11YR48cLqUJb9Cwq2YL3tveVTPeFkZaLP8cdcH5UphVPJ7kYwCUCRLnywd3xkUhb4ZYWtf5-0",
+        );
+        res.unwrap();
```

Comment:
it would be more idiomatic to assert the Result type rather than unwrap and not use the result:
`assert!(res.is_ok());`

Reply:
Fixed at edb54bd09862aca7efc2e1504841d6cec4916173  . I changed it previously to debug a test. IMO, this is a bad sympton of fragile tests. As the negative can come from different underlying reasons. but let's keep aligned with current test structure.